cmd/snap/complete.go

```diff
@@ -0,0 +1,181 @@
+package main
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/client"
+)
+
+type installedSnapName string
+
+func (s installedSnapName) Complete(match string) []flags.Completion {
+	cli := Client()
+	snaps, err := cli.List(nil, nil)
+	if err != nil {
+		return nil
+	}
+
+	ret := make([]flags.Completion, 0, len(snaps))
+	for _, snap := range snaps {
+		if strings.HasPrefix(snap.Name, match) {
+			ret = append(ret, flags.Completion{Item: snap.Name})
+		}
+	}
+
+	return ret
+}
+
+type remoteSnapName string
+
+func (s remoteSnapName) Complete(match string) []flags.Completion {
+	if len(match) < 3 {
+		return nil
+	}
+	cli := Client()
+	snaps, _, err := cli.Find(&client.FindOptions{
+		Prefix: true,
+		Query:  match,
+	})
+	if err != nil {
+		return nil
+	}
+	ret := make([]flags.Completion, len(snaps))
+	for i, snap := range snaps {
+		ret[i] = flags.Completion{Item: snap.Name}
+	}
+	return ret
+}
+
+type snapDir string
+
+func (s snapDir) Complete(match string) []flags.Completion {
+	dir, prefix := filepath.Split(match)
+	if dir == "" {
+		dir = "."
+	}
+	d, err := os.Open(dir)
+	if err != nil {
+		return nil
+	}
+	defer d.Close()
+	var res []flags.Completion
+	for {
+		fis, _ := d.Readdir(100)
```

Comment:
also note the second iteration dropped all these, as file completions are better done in the completion script itself (doing it this way you ended with spaces after each directory)